src/main/java/app/model/agents/WallFollow/WallFollowAgent.java

```diff
@@ -0,0 +1,417 @@
+                (int) (agentCell.getY() + rotateAgentLeft(false).getY()*moveLength),
+                rotateAgentLeft(false).getAngle(), false, 1);
+        cellGraph.updateAgentPos(agentCell,forwardCell,leftCell,rightCell);
+        if (!noMovesDone)
+        {
+            exploreVerticesUntilObstacle(forwardCell, direction, 1);
+            exploreVerticesUntilObstacle(leftCell, rotateAgentLeft(true), 1);
+            exploreVerticesUntilObstacle(rightCell, rotateAgentLeft(false), 1);
+        }
+    }
+
+    public void exploreVerticesUntilObstacle(BooleanCell initVertex, Vector dir, int distFromAgentVertex)
+    {
+        // explore all vertices in a direction until an obstacle is encountered
+        // explore here means agents sees the vertices, not actually moves to them
+        // System.out.println("Checking vertices starting from: " + initVertex.getX() + " " + initVertex.getY());
+        BooleanCell nextCell = getNeighbourVertex((int) (initVertex.getX() + dir.getX()*moveLength),
+                (int) (initVertex.getY() + dir.getY()*moveLength), dir.getAngle(), true, distFromAgentVertex);
+        cellGraph.addExploredVertex(nextCell);
+        if (!nextCell.getObstacle())
+        {
+            //System.out.println("RECURSION level " + distFromAgentVertex);
+            exploreVerticesUntilObstacle(nextCell, dir, distFromAgentVertex+1);
+        }
+        return;
+    }
+
+    public BooleanCell getNeighbourVertex(int neighbourX, int neighbourY, double rayAngle, boolean exploring,
+                                          int distFromAgentVertex)
+    {
+        BooleanCell neighbour;
+        if (!cellGraph.getVertices().containsKey(neighbourX + " " + neighbourY))
+        {
+            neighbour = new BooleanCell(neighbourX, neighbourY);
+            // TODO differentiate between obstacle and occupied cell
+            // TODO write a new method for checking intersection with boundaries
+            if (!exploring && !noWallDetected(rayAngle))
+            {
+                neighbour.setObstacle(true);
+            }
+            if (exploring && !noObstacleDetectedFromPos(rayAngle, distFromAgentVertex))
+            {
+                neighbour.setObstacle(true);
+            }
+        }
+        else
+        {
+            neighbour = cellGraph.getVertices().get(neighbourX + " " + neighbourY);
+            // check if next vertex is initial vertex where started following a wall
+            if (cellGraph.getInitialWallFollowPos() != null &&
+                    rayAngle == direction.getAngle() && !tempAgentCell.equals(cellGraph.getAgentPos())
+                    && neighbour.equals(cellGraph.getInitialWallFollowPos()))
+            {
+                initialVertexFound = true;
+            }
+        }
+        return neighbour;
+    }
+
+    public void updateGraph(Move move)
+    {
+        // if move is rotation
```

Comment:
Consider for all these `//if something` comments; if your If statement is so difficult to read that you need a comment, perhaps it should be it's own function with a better name.  For example
```java
if( Math.sqrt(Math.pow(v.getX) - u.getX()) + Math.pow(v.getY) + u.getY()) > maxDist)
{
   //...
}
```
Could be refactored to:
```java
if(dist(u, v) > maxDist)
```
with a separate function `dist` taking in values and doing the calculation.  This makes the code more readable, and has a secondary effect of making the component parts easier to unit test.